### ISSUE TITLE
RedirectResponse does not set cookies

### ISSUE THREAD

**Issue by natanfelles** [codeigniter4/CodeIgniter4#1393]
**Describe the bug**

In a Controller, the following code works:

```php
return $this->response->redirect(route_to('admin'))
->setCookie('foo', 'bar', YEAR)
->setCookie('login_time', $login_time, YEAR);
```

It does not works:

```php
return redirect()->route('admin')
->setCookie('foo', 'bar', YEAR)
->setCookie('login_time', $login_time, YEAR);
```

**CodeIgniter 4 version**

4.0.0-alpha.2

**Affected module(s)**

HTTP

**Expected behavior, and steps to reproduce if appropriate**

Expected cookies to be set.

**Context**
 - OS: [Debian]
 - Web server [PHP]
 - PHP version [7.2]

**Comment by lonnieezell** [comment]
> `return redirect()->route('admin') ->setCookie('foo', 'bar', YEAR) ->setCookie('login_time', $login_time, YEAR)->send();`
> 
> add send() method is ok. try again.

It shouldn't be necessary to call send there, though. Will look into it.

**Comment by samsonasik** [comment]
I tried to reproduce at #1429 but seems it works with or without send() call.

**Comment by natanfelles** [comment]
Hi, seems not to working, on a fresh install.

**application/Config/development/Routes.php**

```php
$routes->get('redirect/check', 'Redirect::check', ['as' => 'check']);
```

**application/Controllers/Redirect.php**

```php
<?php namespace App\Controllers;

use CodeIgniter\Controller;

class Redirect extends Controller
{
	public function index()
	{
		$login_time = time();

		/*return $this->response->redirect(route_to('check'))
		->setCookie('foo', 'bar', YEAR)
		->setCookie('login_time', $login_time, YEAR);*/

		return redirect()->route('check')
		->setCookie('foo', 'bar', YEAR)
		->setCookie('login_time', $login_time, YEAR);
	}

	public function check()
	{
		var_dump($this->request->getCookie());
	}
}

```

http://localhost:8080/redirect

**Comment by jim-parry** [comment]
I am working on beefier unit testing for HTTP\Response. Will make sure this is addresses & fixed.

**Comment by jim-parry** [comment]
@natanfelles Shouldn't you be checking the response for cookies?

**Comment by natanfelles** [comment]
I think, no.

I check in the request of `check()` because it should be set in the response of the `index()`.

When the commented code run, the cookies are responded. With `redirect()` is not responded.

With `redirect()`:

```
/var/www/codeigniter/application/Controllers/Redirect.php:22:
array (size=4)
  'debug-bar-state' => string 'open' (length=4)
  'debug-bar-tab' => string 'ci-history' (length=10)
```

With `$this->response->redirect`:

```
/var/www/codeigniter/application/Controllers/Redirect.php:22:
array (size=6)
  'debug-bar-state' => string 'open' (length=4)
  'debug-bar-tab' => string 'ci-history' (length=10)
  'foo' => string 'bar' (length=3)
  'login_time' => string '1542303256' (length=10)
```

Note that the time is not updated if you do not clear the browser cookies, when returning with `redirect()`. With the response property the time is always updated, because the `Set-Cookie` header is responded:

Raw headers of Firefox:
```
HTTP/1.1 307 Temporary Redirect
Date: Thu, 15 Nov 2018 17:40:18 GMT
Server: Apache/2.4.29 (Ubuntu)
Cache-control: no-store, max-age=0, no-cache
Location: /redirect/check
Cache-control: no-store, max-age=0, no-cache
Location: /redirect/check
Set-Cookie: foo=bar; expires=Fri, 15-Nov-2019 17:40:18 GMT; Max-Age=31536000; path=/
Set-Cookie: login_time=1542303618; expires=Fri, 15-Nov-2019 17:40:18 GMT; Max-Age=31536000; path=/
Content-Length: 259
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8
```

**Comment by natanfelles** [comment]
Raw header using using the `redirect()` function:

```
HTTP/1.1 307 Temporary Redirect
Date: Thu, 15 Nov 2018 17:49:44 GMT
Server: Apache/2.4.29 (Ubuntu)
Cache-control: no-store, max-age=0, no-cache
Location: http://localhost:8080/redirect/check
Content-Length: 0
Keep-Alive: timeout=5, max=100
Connection: Keep-Alive
Content-Type: text/html; charset=UTF-8
```

**Comment by jim-parry** [comment]
Can finally reproduce. Digging further.

**Comment by jim-parry** [comment]
Hmm - redirect() retuns a RedirectResponse, but $response->redirect() returns a Response.
setCookie() in Response should be used by both.
This is still a mystery - why one would set ookies & the other not.

**Comment by natanfelles** [comment]
When are cookies sent? - just a question, maybe help to found

**Comment by natanfelles** [comment]
I can confirm the problem. Removing the sendHeaders is not redirecting and ending up on a blank page.

I'm going to dedicate myself to this issue now. Coffee is hot.

**Comment by jim-parry** [comment]
I think I solved it with #1504, but this (original) issue may now be back
